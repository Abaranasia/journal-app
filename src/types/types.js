
export const types = {
    /** Este módulo define y resume los tipos para facilitar la identificación de acciones para los dispatchers: */

    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',

    uiStartLoading: '[UI] Start Loading',
    uiFinishLoading: '[UI] Finish Loading',

    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set Active note',
    notesLoad: '[Notes] Load Note',
    notesUpdated: '[Notes] Updated Note',
    notesFileUrl: '[Notes] Updated image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning',
}
