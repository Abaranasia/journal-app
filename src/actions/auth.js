import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";


/** Este módulo define los actions para el proceso de auth.
 *  Aquí incluiríamos todos los métodos de acceso necesarios: email, google, twitter, etc.
 */


export const startLoginEmailPassword = (email, password) => {
    // middleware para login con email que gestiona el action de login

    return (dispatch) => { //devuelve un objeto llamado dispatch con el tipo y el payload

        dispatch(startLoading()); // Empieza el procedimiento de login

        
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
            })
            .catch(e => {
                console.log('Hubo un error en el login:', e);
                dispatch(finishLoading()); // Finaliza el procedimiento de login 
                Swal.fire('Error', e.message, 'error')
            });
/*         setTimeout(() => {
            dispatch(login(123, 'Pedro'))
        }, 3500); */

    }
}

export const startRegisterWithEmailPassword = (email, password, name) => {
    // dado que es ua tarea asíncrona, necesito retornar un callback
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name }); // hacemos esto porque el registro en Firebase no lo hace solo
                console.log(user);

                dispatch(
                    login(user.uid, user.displayName)
                )
            })
            .catch(e => { console.log('Hubo un error al registrarse:', e) });
    }
}

export const startGoogleLogin = () => {
    // middleware para login con usuario de Google que gestiona el action de login

    return (dispatch) => {
    //devuelve una función asíncrona que recibe un dispatch con el tipo y el payload que deben ser enviados

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                //console.log(user.uid, user.displayName);
                dispatch(
                    login(user.uid, user.displayName)
                )
            });
    }
}


export const login = (uid, displayName) => {
    //action para login: devuelve un tipo (login) y un payload con uid y nombre

    return {
        type: types.login,
        payload: {
            uid, displayName
        }
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout())
        dispatch(notesLogout())
    }
}
export const logout = (params) => ({
    type: types.logout
})
