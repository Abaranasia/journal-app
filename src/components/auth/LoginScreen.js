import React from 'react';
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';

export const LoginScreen = () => {

    /** Este módulo tiene acceso a useDispatch e invoca, dependiendo de la vía de acceso,
     * al action correspondiente para cada forma de acceso, ya que alguno requiere middleware */


    const dispatch = useDispatch(); // gestiona el dispatch de redux

    const { loading } = useSelector(state => state.ui) // Esto nos permite identificar si cuándo se está haciendo login

    const [formValues, handleInputChange] = useForm({ //useForm
        email: 'rankirlian@gmail.com',
        password: 'abc123'
    });

    const { email, password } = formValues;


    const handleLogin = (e) => { // para login con email y password
        e.preventDefault();
        //console.log(email, password);
        //dispatch(login(1028201, 'Ran'))
        dispatch(startLoginEmailPassword(email, password)) // recibe type y payload para enviar al dispatcher
    }


    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin()); // recibe type y payload para enviar al dispatcher
    }


    return (
        <>
            <h3 className="auth__title">Login</h3>

            <form onSubmit={handleLogin} className="animate__animated animate__fadeIn animate__faster">
                <input
                    type="text"
                    placeholder="email"
                    name="email"
                    value={email}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="password"
                    name="password"
                    value={password}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary"
                    disabled={loading}
                >
                    Login
                </button>

                <br />
                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to="/auth/register">
                    Create new account
                </Link>
            </form>
        </>
    )
}
