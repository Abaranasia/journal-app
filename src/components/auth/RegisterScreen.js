import React from 'react';
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPassword } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch();

    const { msgError } = useSelector(state => state.ui);
    /** El hook useSelecto permite redirigir info desde el Store a la app
     * Aquí recibe el state de store y lo leemos para poder usar el mensaje de error */
   // console.log(msgError)

    const [formValues, handleInputChange] = useForm({
        name: 'Ran Kirlian',
        email: 'rankirlian@gmail.com',
        password: 'abc123',
        password2: 'abc123'
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            //console.log(name, email, password, password2);
            dispatch(startRegisterWithEmailPassword(email, password, name))
        }
    }

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            //console.log('Name is required');
            return false;
        } //else if (email... contra validator.js) {}
        else if (email.trim().length === 0) {
            dispatch(setError('email is required'));
            //console.log('email is required');
            return false;
        } else if (password !== password2 || password.length < 3) {
            dispatch(setError('password missmatch or malformed'));
            //console.log('password missmatch or malformed');
            return false;
        }
        dispatch(removeError())
        return true
    }


    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form onSubmit={handleRegister} className="animate__animated animate__fadeIn animate__faster">

                {msgError && // Pinta la capa si hay un mensaje de error y en él muestra dicho error
                    (<div className="auth__alert-error">
                        {msgError}
                    </div>)
                }

                <input
                    type="text"
                    placeholder="name"
                    name="name"
                    value={name}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="text"
                    placeholder="email"
                    name="email"
                    value={email}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="password"
                    name="password"
                    value={password}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="confirm password"
                    name="password2"
                    value={password2}
                    className="auth__input"
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary mb-5"
                >
                    Register
                </button>
                <br />
                <Link to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </>
    )
}