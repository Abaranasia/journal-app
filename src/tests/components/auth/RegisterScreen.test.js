import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

/* import { startGoogleLogin, startLoginEmailPassword } from '../../../actions/auth';
 */

/* jest.mock('../../../actions/auth', () => ({
    // Mockeamos esta función porque en el test no pretende evaluarla
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn(),
})
); */


const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Para evaluar estas pruebas vamos a necesitar que nos ayude el store. Para ello empleamos mockStore, 
// MockStore nos permitirá averiguarlas acciones que lo han disparado, con qué argumentos, etc.

const initState = { //Debe tener el estado inicial, el que nos encontraremos antes de hacer login 
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState);
// Nuestro store fictício debe definir aquí el estado inicial en el que necesitamos que esté para estos test

//store.dispatch = jest.fn(); // No necesitamos simular el dispatch porque es todo síncrono


describe('<RegisterScreen> tests', () => {
    /* 
        beforeEach(() => { //Reiniciamos el store porque conserva los estados anteriores
            store = mockStore(initState);
            jest.clearAllMocks
        }) */

    const wrapper = mount(
        <Provider store={store}> {/* lo necesitamos para fingir el context del usuario */}
            <MemoryRouter> {/* Lo necesitamos para que asuma la ruta de react-router */}
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    )


    test('component should render properly', () => {
        expect(wrapper).toMatchSnapshot()
    });


    test('should dispatch the properly action', () => {

        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault() { } // le pasamos la función porque si no se queja. Pasa siempre que sea empleada
        });

        const actions = store.getActions();
        // console.log(actions)
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'email is required'
        })
    });

    test('should show error alert', () => {

        const initState = { //Debe tener el estado inicial, el que nos encontraremos antes de hacer login 
            auth: {},
            ui: {
                loading: false,
                msgError: 'email is required'
            }
        }
        const store = mockStore(initState);

        const wrapper = mount(
            <Provider store={store}> {/* lo necesitamos para fingir el context del usuario */}
                <MemoryRouter> {/* Lo necesitamos para que asuma la ruta de react-router */}
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )

        expect(wrapper.find('.auth__alert-error').exists()).toBe(true)
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)
    });
});