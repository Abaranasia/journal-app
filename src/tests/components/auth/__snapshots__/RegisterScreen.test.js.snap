// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen> tests component should render properly 1`] = `
Array [
  " ",
  Array [
    " ",
    Array [
      <h3
        className="auth__title"
      >
        Register
      </h3>,
      <form
        className="animate__animated animate__fadeIn animate__faster"
        onSubmit={[Function]}
      >
        <input
          className="auth__input"
          name="name"
          onChange={[Function]}
          placeholder="name"
          type="text"
          value="Ran Kirlian"
        />
        <input
          className="auth__input"
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="text"
          value="rankirlian@gmail.com"
        />
        <input
          className="auth__input"
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
          value="abc123"
        />
        <input
          className="auth__input"
          name="password2"
          onChange={[Function]}
          placeholder="confirm password"
          type="password"
          value="abc123"
        />
        <button
          className="btn btn-primary mb-5"
          type="submit"
        >
          Register
        </button>
        <br />
        <a
          href="/auth/login"
          onClick={[Function]}
        >
          Already registered?
        </a>
      </form>,
    ],
  ],
]
`;
